packages/eslint-plugin-udemy/rules/decorator-order/index.js

```diff
@@ -0,0 +1,29 @@
+'use strict';
+
+function orderChecker(context, a, b) {
+    return node => {
+        const decorators = node.decorators;
+        if (!decorators) { return; }
+        const docoratorNames = decorators.map(d => d.expression.name);
+        const ai = docoratorNames.indexOf(a);
+        const bi = docoratorNames.indexOf(b);
+        if (ai >= 0 && bi >= 0 && bi < ai) {
+            context.report({
+                node,
+                loc: decorators[ai].loc,
+                message: `\`@${a}\` should come before \`@${b}\`.`,
+            });
+        }
+    };
+}
+
+module.exports.rules = {
+    'decorator-order': {
+        create(context) {
+            return {
+                MethodDefinition: orderChecker(context, 'autobind', 'action'),
+                ClassDeclaration: orderChecker(context, 'inject', 'observer'),
```

Comment:
Should the logic be `@observer should be the first decorator`, rather than `@observer should be before @inject`?

I thought in general `@observer` should be first: https://udemy.slack.com/archives/C04HUGFL0/p1530223702000052?thread_ts=1530219678.000357&cid=C04HUGFL0

There could be other decorators besides `@inject`. For example, in the Slack thread, Carmen was using `@deviceType`.

Reply:
This can also get confusing as to what we mean by "first".  In the AST there is simply an array of decorators ordered how they are written left to right. I think by "first" you mean inner-most, which would be last in the array. 

I was writing the rule with another thread in mind where people were mostly concerned with the order of decorators relative to one another. I will change my check to make sure that if `@observer` is there that it is last in the array. But does the same go for `@action` or is it that `@autobind` needs to go first, or both?

Reply:
My bad- yes, by "first" I mean "inner-most".

As far as I'm aware, the only problem we've seen regarding the `@action` decorator is that is that `@action @autobind` doesn't work, so I think what you currently have is fine. I think enforcing `@action` to be innermost is probably fine as well, but I'm not 100% sure about it.